### PR TITLE
fix: progressImpl properly middleware tasks to set the proper result status

### LINKED ISSUE #4341: ProgressImpl: impossible to fail a task
### Bug description

The following lines indicates, that the task created with the `withWidget` can never reach the `catch` handle since an await is located on the `then` handle. Thus preventing from failing a task properly.

https://github.com/containers/podman-desktop/blob/d0cafacb0c92300897b03a6e97586c8fcb6b71c4/packages/main/src/plugin/progress-impl.ts#L94-L105

### Operating system

Windows 11

### Installation Method

Installer from website/GitHub releases

### Version

next (development version)

### Steps to reproduce

Create a simple progress task and try to fail it.

```
progressImpl.withProgress<void>(
  { location: ProgressLocation.TASK_WIDGET, title: `Pulling ${imageName}.` },
  async  progress => {
    throw new Error('error');
});
```

### Relevant log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by axel7083**
### What does this PR do?

Fixes #4341

### CODE REVIEW

File: packages/main/src/plugin/progress-impl.ts [hunk 1]
```diff
@@ -91,18 +91,22 @@ export class ProgressImpl {
...
 
-    task_.catch((err: unknown) => {
-      t.status = 'failure';
-      t.state = 'completed';
-      t.error = String(err);
-      this.taskManager.updateTask(t);
-    });
+    await Promise.any([
+      task_.then(async () => {
+        console.log('progressImpl THEN');
```
Comment (benoitf) [inline packages/main/src/plugin/progress-impl.ts]: looks like you forgot to remove traces

File: packages/main/src/plugin/progress-impl.ts [hunk 2]
```diff
@@ -91,18 +91,22 @@ export class ProgressImpl {
...
+        t.status = 'success';
+        t.state = 'completed';
+        this.taskManager.updateTask(t);
+      }),
+      task_.catch((err: unknown) => {
+        console.log('progressImpl on failure');
```
Comment (benoitf) [inline packages/main/src/plugin/progress-impl.ts]: trace to remove

**Review by benoitf — COMMENTED** [review]
thanks @axel7083 could you add a unit test to ensure that the current code is failing and with the updated code we're reaching the expected behaviour ?

**Review by jeffmaury — COMMENTED** [review]
Looks barely readable to me.
Could be rewritten either through method chaining: task.`then(...).catch(....)` or as:

```
try {
  await task_;
  t.status = 'success';
} catch (err) {
  t.status = 'failure';
  t.error = String(err);
} finally {
  t.state = 'completed';
  this.taskManager.updateTask(t);
}
```

**Comment by axel7083** [comment]
> Looks barely readable to me. Could be rewritten either through method chaining: task.`then(...).catch(....)` or as:
> 
> ```
> try {
>   await task_;
>   t.status = 'success';
> } catch (err) {
>   t.status = 'failure';
>   t.error = String(err);
> } finally {
>   t.state = 'completed';
>   this.taskManager.updateTask(t);
> }
> ```

The return type of `async withWidget<R>(...)` is `Promise<R>`, if we wait for the completion of the task don't we prevent caller from handling the exception of the task ?

I added a unit test to ensure two important element
- The error is not handled by the withWidged, it is propagated to the caller
- The state of the task is set to **completed** and status to **failure**

I changed the implementation to return a new Promise<R>

**Comment by benoitf** [comment]
@jeffmaury could you check again ?

File: packages/main/src/plugin/progress-impl.ts [hunk 3]
```diff
@@ -77,32 +77,35 @@ export class ProgressImpl {
...
-      t.state = 'completed';
-      this.taskManager.updateTask(t);
-    });
 
-    task_.catch((err: unknown) => {
-      t.status = 'failure';
-      t.state = 'completed';
-      t.error = String(err);
-      this.taskManager.updateTask(t);
+    return new Promise<R>((resolve, reject) => {
```
Comment (jeffmaury) [inline packages/main/src/plugin/progress-impl.ts]: As `then`, `catch` and `finally` methods allow methods chaining, I think there is an extra Promise object created that is not required.

Comment (axel7083) [reply packages/main/src/plugin/progress-impl.ts]: I remove the encapsulating promise ! Thanks for noticing

**Comment by axel7083** [comment]
CC @jeffmaury could you take a look ?

**Review by jeffmaury — APPROVED** [review]
LGTM

**Pull request merged**